modules/calc/src/main/java/com/opengamma/strata/calc/runner/CalculationFunctions.java

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.calc.runner;
+
+import static com.opengamma.strata.collect.Guavate.toImmutableMap;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.ImmutableValidator;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaBean;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.light.LightMetaBean;
+
+import com.google.common.collect.ImmutableMap;
+import com.opengamma.strata.basics.CalculationTarget;
+import com.opengamma.strata.calc.runner.function.CalculationFunction;
+import com.opengamma.strata.collect.ArgChecker;
+
+/**
+ * The calculation functions.
+ * <p>
+ * This provides the complete set of functions that will be used in a calculation.
+ * Each {@link CalculationFunction} handles a specific type of {@link CalculationTarget},
+ * thus the functions are keyed in a {@code Map} by the target type {@code Class}.
+ */
+@BeanDefinition(style = "light")
+public final class CalculationFunctions implements ImmutableBean, Serializable {
```

Comment:
The `getFunction` and `findFunction` methods have a parameter for the target, not the target type. I think that's a very good thing as it leaves the door open to choosing the function based on more complicated criteria. For example it would be possible to have different functions for different types of swap where the target type is always `Swap` but the legs have different types. It would also allow target attributes to be used when choosing the function.

The only problem is that this class is a concrete final type and the methods for finding functions are not defined on an interface. Which means this is the only possible implementation.

Could we convert `CalculationFunctions` to an interface containing `getFunction` and `findFunction` and rename this class? Maybe `TargetTypeCalculationFunctions`.